misc/priority-inversion.md

```diff
@@ -1 +1,251 @@
+preemptible&mdash; developers must choose to explicitly yield or interact with
+a yieldy API. Assessing the risk here is also difficult since the priority can
+be lost when [mixing async APIs](userspace-task-models.md#mixing-async-apis)
+with `scheduler.postTask()`.
+
+There may be an opportunity for a UA to detect and mitigate priority inversions
+that involve primitives like [web
+locks](https://developer.mozilla.org/en-US/docs/Web/API/Web_Locks_API), but
+that likely requires further refinement of the userspace task model and
+priority propagation, which is out of scope for `scheduler.postTask()`. We
+believe the best approach for preventing this type of priority inversion is to
+be clear that developers need to be careful not to keep resources held while
+yielding.
+
+## Other Forms of Priority Inversion
+
+Priority inversions can also occur if there is a dependency between a
+not-yet-run low priority task and a higher priority task. This section gives a
+couple of examples of how this can occur, along with application-level
+mitigations.  These examples are not meant to be exhaustive, but we feel they
+help illustrate the types of priority inversions that developers should be
+aware of.
+
+### Example 1: Non-Locked Shared Resources
+
+A significantly complex app may have a system for retrieving and caching
+resources that is used by multiple parts of the page to avoid duplication of
+work. Furthermore, the system might use priorities to schedule work based on
+the impact on user experience.
+
+The risk of priority inversion occurs here if a
+resource request is started at low priority and a subsequent high priority
+request is initiated while the low priority one is pending:
+
+```js
+// Note: we assume parts of the processing for such a system is done with
+// prioritized tasks. This example would also apply to fetch() if the API was
+// aware of postTask priorities.
+function getCachedResource(resource, priority) {
+  if (needToFetchResource(resource) {
+    // Kick off the request in a separate task of the appropriate priority.
+    let result = scheduler.postTask(() => {
+      return getResourceAtPriority(resource, priority);
+    }, {priority});
+
+    // Save off the resulting Promise, which will signal completion.
+    setPendingResource(result);
+  }
+
+  // Return a promise that will be resolved with the data --- either when the
+  // fetching task completes or now if we already have the data.
+  return getPromiseForResourceData(resource);
+}
+
+...
+
+// The first request for |url| happens at background priority.
+getCachedResource(url, 'background').then(...);
+
+...
+
+// A subsequent request for |url| happens just after, at user-blocking priority
+// This can lead to a priority inversion if the system is not designed to
+// handle this case.
+getCachedResource(url, 'user-blocking').then(...);
+```
+
+Developers should be aware that such situations could arise. Systems like this
```

Comment:
remove "like this"? It's probably clearer without.

Reply:
sgtm, done.